In [57]:
import sqlite3
import pandas as pd
from sklearn.tree import DecisionTreeRegressor
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from math import sqrt

In [58]:
cnx = sqlite3.connect('database.sqlite')
df = pd.read_sql_query("SELECT * FROM Player_Attributes", cnx)

In [59]:
df.head()

id  player_fifa_api_id  player_api_id                 date  overall_rating  \
0   1              218353         505942  2016-02-18 00:00:00            67.0   
1   2              218353         505942  2015-11-19 00:00:00            67.0   
2   3              218353         505942  2015-09-21 00:00:00            62.0   
3   4              218353         505942  2015-03-20 00:00:00            61.0   
4   5              218353         505942  2007-02-22 00:00:00            61.0   

   potential preferred_foot attacking_work_rate defensive_work_rate  crossing  \
0       71.0          right              medium              medium      49.0   
1       71.0          right              medium              medium      49.0   
2       66.0          right              medium              medium      49.0   
3       65.0          right              medium              medium      48.0   
4       65.0          right              medium              medium      48.0   

      ...       vision  penalties  marking  standing_tackle  sliding_tackle  \
0     ...         54.0       48.0     65.0             69.0            69.0   
1     ...         54.0       48.0     65.0             69.0            69.0   
2     ...         54.0       48.0     65.0             66.0            69.0   
3     ...         53.0       47.0     62.0             63.0            66.0   
4     ...         53.0       47.0     62.0             63.0            66.0   

   gk_diving  gk_handling  gk_kicking  gk_positioning  gk_reflexes  
0        6.0         11.0        10.0             8.0          8.0  
1        6.0         11.0        10.0             8.0          8.0  
2        6.0         11.0        10.0             8.0          8.0  
3        5.0         10.0         9.0             7.0          7.0  
4        5.0         10.0         9.0             7.0          7.0  

[5 rows x 42 columns]

In [60]:
df.columns

Index(['id', 'player_fifa_api_id', 'player_api_id', 'date', 'overall_rating',
       'potential', 'preferred_foot', 'attacking_work_rate',
       'defensive_work_rate', 'crossing', 'finishing', 'heading_accuracy',
       'short_passing', 'volleys', 'dribbling', 'curve', 'free_kick_accuracy',
       'long_passing', 'ball_control', 'acceleration', 'sprint_speed',
       'agility', 'reactions', 'balance', 'shot_power', 'jumping', 'stamina',
       'strength', 'long_shots', 'aggression', 'interceptions', 'positioning',
       'vision', 'penalties', 'marking', 'standing_tackle', 'sliding_tackle',
       'gk_diving', 'gk_handling', 'gk_kicking', 'gk_positioning',
       'gk_reflexes'],
      dtype='object')

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 183978 entries, 0 to 183977
Data columns (total 42 columns):
id                     183978 non-null int64
player_fifa_api_id     183978 non-null int64
player_api_id          183978 non-null int64
date                   183978 non-null object
overall_rating         183142 non-null float64
potential              183142 non-null float64
preferred_foot         183142 non-null object
attacking_work_rate    180748 non-null object
defensive_work_rate    183142 non-null object
crossing               183142 non-null float64
finishing              183142 non-null float64
heading_accuracy       183142 non-null float64
short_passing          183142 non-null float64
volleys                181265 non-null float64
dribbling              183142 non-null float64
curve                  181265 non-null float64
free_kick_accuracy     183142 non-null float64
long_passing           183142 non-null float64
ball_control           183142 non-null float64
accele

In [62]:
df.describe(include='all')

id  player_fifa_api_id  player_api_id                 date  \
count   183978.00000       183978.000000  183978.000000               183978   
unique           NaN                 NaN            NaN                  197   
top              NaN                 NaN            NaN  2007-02-22 00:00:00   
freq             NaN                 NaN            NaN                11794   
mean     91989.50000       165671.524291  135900.617324                  NaN   
std      53110.01825        53851.094769  136927.840510                  NaN   
min          1.00000            2.000000    2625.000000                  NaN   
25%      45995.25000       155798.000000   34763.000000                  NaN   
50%      91989.50000       183488.000000   77741.000000                  NaN   
75%     137983.75000       199848.000000  191080.000000                  NaN   
max     183978.00000       234141.000000  750584.000000                  NaN   

        overall_rating      potential preferred_foot attacking_work_rate  \
count    183142.000000  183142.000000         183142              180748   
unique             NaN            NaN              2                   8   
top                NaN            NaN          right              medium   
freq               NaN            NaN         138409              125070   
mean         68.600015      73.460353            NaN                 NaN   
std           7.041139       6.592271            NaN                 NaN   
min          33.000000      39.000000            NaN                 NaN   
25%          64.000000      69.000000            NaN                 NaN   
50%          69.000000      74.000000            NaN                 NaN   
75%          73.000000      78.000000            NaN                 NaN   
max          94.000000      97.000000            NaN                 NaN   

       defensive_work_rate       crossing      ...               vision  \
count               183142  183142.000000      ...        181265.000000   
unique                  19            NaN      ...                  NaN   
top                 medium            NaN      ...                  NaN   
freq                130846            NaN      ...                  NaN   
mean                   NaN      55.086883      ...            57.873550   
std                    NaN      17.242135      ...            15.144086   
min                    NaN       1.000000      ...             1.000000   
25%                    NaN      45.000000      ...            49.000000   
50%                    NaN      59.000000      ...            60.000000   
75%                    NaN      68.000000      ...            69.000000   
max                    NaN      95.000000      ...            97.000000   

            penalties        marking  standing_tackle  sliding_tackle  \
count   183142.000000  183142.000000    183142.000000   181265.000000   
unique            NaN            NaN              NaN             NaN   
top               NaN            NaN              NaN             NaN   
freq              NaN            NaN              NaN             NaN   
mean        55.003986      46.772242        50.351257       48.001462   
std         15.546519      21.227667        21.483706       21.598778   
min          2.000000       1.000000         1.000000        2.000000   
25%         45.000000      25.000000        29.000000       25.000000   
50%         57.000000      50.000000        56.000000       53.000000   
75%         67.000000      66.000000        69.000000       67.000000   
max         96.000000      96.000000        95.000000       95.000000   

            gk_diving    gk_handling     gk_kicking  gk_positioning  \
count   183142.000000  183142.000000  183142.000000   183142.000000   
unique            NaN            NaN            NaN             NaN   
top               NaN            NaN            NaN             NaN   
freq              NaN            NaN            NaN             NaN   
mean       

In [63]:
df.shape

(183978, 42)

In [64]:
df.isna().sum()

id                        0
player_fifa_api_id        0
player_api_id             0
date                      0
overall_rating          836
potential               836
preferred_foot          836
attacking_work_rate    3230
defensive_work_rate     836
crossing                836
finishing               836
heading_accuracy        836
short_passing           836
volleys                2713
dribbling               836
curve                  2713
free_kick_accuracy      836
long_passing            836
ball_control            836
acceleration            836
sprint_speed            836
agility                2713
reactions               836
balance                2713
shot_power              836
jumping                2713
stamina                 836
strength                836
long_shots              836
aggression              836
interceptions           836
positioning             836
vision                 2713
penalties               836
marking                 836
standing_tackle     

In [65]:
df.duplicated().any()

False

In [66]:
df = df.drop(['id','player_fifa_api_id','player_api_id','date'], axis=1)

In [67]:
df.columns

Index(['overall_rating', 'potential', 'preferred_foot', 'attacking_work_rate',
       'defensive_work_rate', 'crossing', 'finishing', 'heading_accuracy',
       'short_passing', 'volleys', 'dribbling', 'curve', 'free_kick_accuracy',
       'long_passing', 'ball_control', 'acceleration', 'sprint_speed',
       'agility', 'reactions', 'balance', 'shot_power', 'jumping', 'stamina',
       'strength', 'long_shots', 'aggression', 'interceptions', 'positioning',
       'vision', 'penalties', 'marking', 'standing_tackle', 'sliding_tackle',
       'gk_diving', 'gk_handling', 'gk_kicking', 'gk_positioning',
       'gk_reflexes'],
      dtype='object')

In [68]:
df.dropna(subset=['overall_rating', 'potential', 'preferred_foot', 'attacking_work_rate',
       'defensive_work_rate', 'crossing', 'finishing', 'heading_accuracy',
       'short_passing', 'volleys', 'dribbling', 'curve', 'free_kick_accuracy',
       'long_passing', 'ball_control', 'acceleration', 'sprint_speed',
       'agility', 'reactions', 'balance', 'shot_power', 'jumping', 'stamina',
       'strength', 'long_shots', 'aggression', 'interceptions', 'positioning',
       'vision', 'penalties', 'marking', 'standing_tackle', 'sliding_tackle',
       'gk_diving', 'gk_handling', 'gk_kicking', 'gk_positioning',
       'gk_reflexes'], how='all', inplace=True)

In [69]:
df.isna().sum()

overall_rating            0
potential                 0
preferred_foot            0
attacking_work_rate    2394
defensive_work_rate       0
crossing                  0
finishing                 0
heading_accuracy          0
short_passing             0
volleys                1877
dribbling                 0
curve                  1877
free_kick_accuracy        0
long_passing              0
ball_control              0
acceleration              0
sprint_speed              0
agility                1877
reactions                 0
balance                1877
shot_power                0
jumping                1877
stamina                   0
strength                  0
long_shots                0
aggression                0
interceptions             0
positioning               0
vision                 1877
penalties                 0
marking                   0
standing_tackle           0
sliding_tackle         1877
gk_diving                 0
gk_handling               0
gk_kicking          

In [70]:
df['attacking_work_rate'].value_counts()

medium    125070
high       42823
low         8569
None        3639
norm         348
y            106
le           104
stoc          89
Name: attacking_work_rate, dtype: int64

In [71]:
df['attacking_work_rate'].unique()

array(['medium', 'high', None, 'low', 'None', 'le', 'norm', 'stoc', 'y'],
      dtype=object)

In [72]:
df['attacking_work_rate'].fillna(method='ffill',inplace = True, axis = 0)

In [73]:
df['attacking_work_rate'].value_counts()

medium    126858
high       43214
low         8659
None        3742
norm         360
le           112
y            108
stoc          89
Name: attacking_work_rate, dtype: int64

In [74]:
df['attacking_work_rate'].unique()

array(['medium', 'high', 'low', 'None', 'le', 'norm', 'stoc', 'y'],
      dtype=object)

In [75]:
attacking_work_rate_dummies = pd.get_dummies(df['attacking_work_rate'],prefix='attacking_work_rate').iloc[:, 1:]
attacking_work_rate_dummies.head(2)

attacking_work_rate_high  attacking_work_rate_le  attacking_work_rate_low  \
0                         0                       0                        0   
1                         0                       0                        0   

   attacking_work_rate_medium  attacking_work_rate_norm  \
0                           1                         0   
1                           1                         0   

   attacking_work_rate_stoc  attacking_work_rate_y  
0                         0                      0  
1                         0                      0

In [76]:
df['volleys'].fillna((df['volleys'].mean()), inplace=True)
df['volleys'].isna().sum()

0

In [77]:
df['curve'].fillna((df['curve'].mean()), inplace=True)
df['curve'].isna().sum()

0

In [78]:
df['agility'].fillna((df['curve'].mean()), inplace=True)
df['agility'].isna().sum()

0

In [79]:
df['balance'].fillna((df['balance'].mean()), inplace=True)
df['balance'].isna().sum()

0

In [80]:
df['jumping'].fillna((df['jumping'].mean()), inplace=True)
df['jumping'].isna().sum()

0

In [81]:
df['vision'].fillna((df['vision'].mean()), inplace=True)
df['vision'].isna().sum()

0

In [82]:
df['sliding_tackle'].fillna((df['sliding_tackle'].mean()), inplace=True)
df['sliding_tackle'].isna().sum()

0

In [83]:
df['overall_rating'].unique()

array([67., 62., 61., 74., 73., 71., 70., 69., 68., 65., 64., 54., 51.,
       52., 47., 53., 66., 59., 75., 72., 76., 78., 77., 79., 60., 80.,
       81., 82., 84., 48., 63., 83., 55., 58., 50., 56., 49., 57., 42.,
       46., 45., 85., 44., 86., 89., 87., 88., 91., 40., 90., 41., 43.,
       38., 93., 92., 39., 33., 36., 37., 35., 94.])

In [84]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 183142 entries, 0 to 183977
Data columns (total 38 columns):
overall_rating         183142 non-null float64
potential              183142 non-null float64
preferred_foot         183142 non-null object
attacking_work_rate    183142 non-null object
defensive_work_rate    183142 non-null object
crossing               183142 non-null float64
finishing              183142 non-null float64
heading_accuracy       183142 non-null float64
short_passing          183142 non-null float64
volleys                183142 non-null float64
dribbling              183142 non-null float64
curve                  183142 non-null float64
free_kick_accuracy     183142 non-null float64
long_passing           183142 non-null float64
ball_control           183142 non-null float64
acceleration           183142 non-null float64
sprint_speed           183142 non-null float64
agility                183142 non-null float64
reactions              183142 non-null float64

In [85]:
preferred_foot_dummies = pd.get_dummies(df['preferred_foot'],prefix='preferred_foot')
preferred_foot_dummies.head(2)

preferred_foot_left  preferred_foot_right
0                    0                     1
1                    0                     1

In [86]:
df['defensive_work_rate'].value_counts()

medium    130846
high       27041
low        18432
_0          2394
o           1550
1            441
ormal        348
2            342
3            258
5            234
7            217
0            197
6            197
9            152
4            116
es           106
ean          104
tocky         89
8             78
Name: defensive_work_rate, dtype: int64

In [87]:
defensive_work_rate_dummies = pd.get_dummies(df['defensive_work_rate'],prefix='defensive_work_rate').iloc[:, 1:]
defensive_work_rate_dummies.head(2)

defensive_work_rate_1  defensive_work_rate_2  defensive_work_rate_3  \
0                      0                      0                      0   
1                      0                      0                      0   

   defensive_work_rate_4  defensive_work_rate_5  defensive_work_rate_6  \
0                      0                      0                      0   
1                      0                      0                      0   

   defensive_work_rate_7  defensive_work_rate_8  defensive_work_rate_9  \
0                      0                      0                      0   
1                      0                      0                      0   

   defensive_work_rate__0  defensive_work_rate_ean  defensive_work_rate_es  \
0                       0                        0                       0   
1                       0                        0                       0   

   defensive_work_rate_high  defensive_work_rate_low  \
0                         0                        0   
1                         0                        0   

   defensive_work_rate_medium  defensive_work_rate_o  \
0                           1                      0   
1                           1                      0   

   defensive_work_rate_ormal  defensive_work_rate_tocky  
0                          0                          0  
1                          0                          0

In [88]:
attacking_work_rate_dummies = pd.get_dummies(df['attacking_work_rate'],prefix='attacking_work_rate').iloc[:, 1:]
attacking_work_rate_dummies.head(2)

attacking_work_rate_high  attacking_work_rate_le  attacking_work_rate_low  \
0                         0                       0                        0   
1                         0                       0                        0   

   attacking_work_rate_medium  attacking_work_rate_norm  \
0                           1                         0   
1                           1                         0   

   attacking_work_rate_stoc  attacking_work_rate_y  
0                         0                      0  
1                         0                      0

In [89]:
df = df.drop(['preferred_foot','attacking_work_rate','defensive_work_rate'],axis=1)

In [90]:
df['overall_rating'].unique()

array([67., 62., 61., 74., 73., 71., 70., 69., 68., 65., 64., 54., 51.,
       52., 47., 53., 66., 59., 75., 72., 76., 78., 77., 79., 60., 80.,
       81., 82., 84., 48., 63., 83., 55., 58., 50., 56., 49., 57., 42.,
       46., 45., 85., 44., 86., 89., 87., 88., 91., 40., 90., 41., 43.,
       38., 93., 92., 39., 33., 36., 37., 35., 94.])

In [91]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 183142 entries, 0 to 183977
Data columns (total 35 columns):
overall_rating        183142 non-null float64
potential             183142 non-null float64
crossing              183142 non-null float64
finishing             183142 non-null float64
heading_accuracy      183142 non-null float64
short_passing         183142 non-null float64
volleys               183142 non-null float64
dribbling             183142 non-null float64
curve                 183142 non-null float64
free_kick_accuracy    183142 non-null float64
long_passing          183142 non-null float64
ball_control          183142 non-null float64
acceleration          183142 non-null float64
sprint_speed          183142 non-null float64
agility               183142 non-null float64
reactions             183142 non-null float64
balance               183142 non-null float64
shot_power            183142 non-null float64
jumping               183142 non-null float64
stamina        

In [92]:
df.isna().sum().any()

False

In [93]:
# Linear Regression

In [94]:
import matplotlib.pyplot as plt
%matplotlib inline

In [96]:
df.columns

Index(['overall_rating', 'potential', 'crossing', 'finishing',
       'heading_accuracy', 'short_passing', 'volleys', 'dribbling', 'curve',
       'free_kick_accuracy', 'long_passing', 'ball_control', 'acceleration',
       'sprint_speed', 'agility', 'reactions', 'balance', 'shot_power',
       'jumping', 'stamina', 'strength', 'long_shots', 'aggression',
       'interceptions', 'positioning', 'vision', 'penalties', 'marking',
       'standing_tackle', 'sliding_tackle', 'gk_diving', 'gk_handling',
       'gk_kicking', 'gk_positioning', 'gk_reflexes'],
      dtype='object')

In [107]:
y = df['overall_rating']
x = df.drop(['overall_rating'],axis=1)

In [100]:
# evaluate the model by splitting into train and test sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=0)

In [101]:
x_train.shape, x_test.shape, y_test.shape, y_test.shape

((128199, 34), (54943, 34), (54943,), (54943,))

In [102]:
model2 = LinearRegression()
model2.fit(x_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [103]:
# predict class labels for the test set
predicted = model2.predict(x_test)
predicted

array([74.64720228, 52.30474369, 81.17642934, ..., 72.64576885,
       70.92014914, 69.88461176])

In [104]:
model2.score(x_test,y_test)

0.8401981293510307


84.29% Accuracy


Model Evaluation Using Cross-Validation¶


In [105]:
# evaluate the model using 10-fold cross-validation
from sklearn.model_selection import cross_val_score
scores = cross_val_score(LinearRegression(), x, y, cv=10)
scores, scores.mean()

(array([0.84610838, 0.84048506, 0.83439035, 0.83766995, 0.84126038,
        0.83923133, 0.84807995, 0.83232622, 0.84166535, 0.84302767]),
 0.8404244617826195)


84.26% Accuracy



In all 3 model evaluation, accuracy is around 84% which means data is neither overfit and accuracy seems decent.
